File: app/src/main/java/com/micewine/emu/core/WineWrapper.java
package com.micewine.emu.core;

import static com.micewine.emu.activities.MainActivity.deviceArch;
import static com.micewine.emu.activities.MainActivity.usrDir;
import static com.micewine.emu.activities.MainActivity.wineDisksFolder;
import static com.micewine.emu.activities.MainActivity.winePrefix;
import static com.micewine.emu.activities.MainActivity.winePrefixesDir;
import static com.micewine.emu.core.EnvVars.getEnv;
import static com.micewine.emu.core.ShellLoader.runCommand;
import static com.micewine.emu.core.ShellLoader.runCommandWithOutput;
import static com.micewine.emu.fragments.DebugSettingsFragment.availableCPUs;

import android.system.ErrnoException;
import android.system.Os;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.List;

public class WineWrapper {
    private static final String IS_BOX64 = deviceArch.equals("x86_64") ? "" : "box64";

    public static String getCpuHexMask(String cpuAffinityMask) {
        int availCpus = Runtime.getRuntime().availableProcessors();
        List<Character> cpuMask = new ArrayList<>();

        for (int i = 0; i < availCpus; i++) {
            cpuMask.add('0');
        }

        String cpuAffinity = cpuAffinityMask.replace(",", "");

        for (char element : cpuAffinity.toCharArray()) {
            int index = Math.abs(Character.getNumericValue(element) - availCpus) - 1;
            cpuMask.set(index, '1');
        }

        StringBuilder binary = new StringBuilder();
        for (char c : cpuMask) {
            binary.append(c);
        }

        return Integer.toHexString(Integer.parseInt(binary.toString(), 2));
    }

    public static boolean[] maskToCpuAffinity(long mask) {
        int availCpus = Runtime.getRuntime().availableProcessors();
        boolean[] affinity = new boolean[availCpus];

        for (int i = 0; i < availCpus; i++) {
            if (((mask >> i) & 1L) == 1L) {
                affinity[i] = true;
            }
        }

        return affinity;
    }

    public static void waitForProcess(String name) {
        while (true) {
            List<ExeProcess> exeProcesses = getExeProcesses();
            for (ExeProcess exeProcess : exeProcesses) {
                if (exeProcess.name.equals(name)) {
                    return;
                }
            }
            try {
                Thread.sleep(125);
            } catch (InterruptedException ignored) {
            }
        }
    }

    public static void wine(String args) {
        wine(args, null);
    }

    public static void wine(String args, String cwd) {
        runCommand(
                ((cwd != null) ? "cd " + cwd + ";" : "") + getEnv() + "WINEPREFIX='" + winePrefixesDir + "/" + winePrefix + "' " + IS_BOX64 + " wine " + args , true
        );
    }

    public static void killAll() {
        runCommand(getEnv() + "WINEPREFIX='" + winePrefixesDir + "/" + winePrefix + "' " + IS_BOX64 + " wineserver -k", false);
        runCommand("pkill -SIGINT -f .exe", false);
        runCommand("pkill -SIGINT -f wineserver", false);
    }

    public static void clearDrives() {
        char letter = 'e';

        while (letter <= 'y') {
            File diskFile = new File(wineDisksFolder + "/" + letter + ":");
            if (diskFile.exists()) {
                diskFile.delete();
            }
            letter++;
        }
    }

    public static void addDrive(String path) {
        runCommand("ln -sf " + path + " " + wineDisksFolder + "/" + getAvailableDisks().get(0) + ":", false);
    }

    private static List<String> getAvailableDisks() {
        char letter = 'e';

        List<String> availableDisks = new ArrayList<>();

        while (letter <= 'z') {
            File diskFile = new File(wineDisksFolder + "/" + letter + ":");
            if (!diskFile.exists()) {
                availableDisks.add(String.valueOf(letter));
            }
            letter++;
        }

        return availableDisks;
    }

    public static void extractIcon(String exePath, String output) {
        if (exePath.toLowerCase().endsWith(".exe")) {
            runCommand(
                    getEnv() + "wrestool -x -t 14 '" + getSanitizedPath(exePath) + "' > '" + output + "'", false
            );
        }
    }

    public static String getSanitizedPath(String filePath) {
        return filePath.replace("'", "'\\''");
    }

    private static String getProcessPath(String processName, String processCwd) {
        File file = new File(processCwd, processName);
        if (file.exists()) {
            return file.toString();
        }

        String[] searchPaths = {
                wineDisksFolder + "/c:/windows/system32",
                wineDisksFolder + "/c:/windows/",
        };

        for (String path : searchPaths) {
            File exeFile = new File(path, processName);
            if (exeFile.exists()) {
                return path + "/" + processName;
            }
        }

        return "";
    }

    private static int getProcessRamUsageKB(int pid) {
        File statusFile = new File("/proc/" + pid + "/status");
        if (!statusFile.exists()) return 0;

        try {
            List<String> lines = Files.readAllLines(statusFile.toPath());

            for (String line : lines) {
                if (line.startsWith("VmRSS:")) {
                    return Integer.parseInt(line.substring(line.indexOf(":") + 1).replace("kB", "").trim());
                }
            }
        } catch (IOException ignored) {
        }

        return 0;
    }

    public static String getProcessCPUAffinity(int pid) {
        File statusFile = new File("/proc/" + pid + "/status");
        if (!statusFile.exists()) return "0";

        try {
            List<String> lines = Files.readAllLines(statusFile.toPath());

            for (String line : lines) {
                if (line.startsWith("Cpus_allowed:")) {
                    return line.substring(line.indexOf(":") + 1).replace(",", "").trim();
                }
            }
        } catch (IOException ignored) {
        }

        return "0";
    }

    public static int getWinPidByIndex(int index) {
        if (index < 0) return -1;

        String[] taskList = runCommandWithOutput(
                getEnv() + "WINEDEBUG=0 BOX64_LOG=0 WINEPREFIX='" + winePrefixesDir + "/" + winePrefix + "' " + IS_BOX64 + " wine tasklist | grep .exe", false
        ).split("\n");

        if (index > taskList.length) return -1;

        return Integer.parseInt(taskList[index].replaceAll(" +", " ").split(" ")[1]);
    }

    public static List<ExeProcess> getExeProcesses() {
        List<ExeProcess> exeProcesses = new ArrayList<>();

        File[] processes = new File("/proc").listFiles();

        if (processes == null) return exeProcesses;

        for (File processFile : processes) {
            if (processFile.isDirectory()) {
                int unixPid;

                try {
                    unixPid = Integer.parseInt(processFile.getName());
                } catch (NumberFormatException ignored) {
                    continue;
                }

                File cmdlineFile = new File(processFile, "cmdline");
                if (!cmdlineFile.exists()) continue;

                try {
                    String cmdline = Files.readAllLines(cmdlineFile.toPath()).toString().trim();
                    String processName = cmdline.split("\u0000")[0].substring(1);

                    if (!processName.toLowerCase().endsWith(".exe")) continue;

                    processName = processName.substring(processName.lastIndexOf("\\") + 1);

                    String cwd;

                    try {
                        cwd = Os.readlink(processFile.toPath() + "/cwd");
                    } catch (ErrnoException ignored) {
                        cwd = "/";
                    }

                    String processPath = getProcessPath(processName, cwd);
                    String iconPath = usrDir + "/icons/" + processName.substring(0, processName.indexOf(".exe")) + "-thumbnail";
                    int ramUsageKB = getProcessRamUsageKB(unixPid);
                    float cpuUsage;

                    try {
                        cpuUsage = Float.parseFloat(runCommandWithOutput("ps -p " + unixPid + " -o %cpu=", false).trim());
                    } catch (NumberFormatException ignored) {
                        cpuUsage = 0F;
                    }

                    extractIcon(processPath, iconPath);

                    exeProcesses.add(
                            new ExeProcess(processName, unixPid, cwd, processPath, iconPath, ramUsageKB, cpuUsage / availableCPUs.length)
                    );
                } catch (IOException ignored) {
                }
            }
        }

        return exeProcesses;
    }

    public static class ExeProcess {
        String name;
        int unixPid;
        String cwd;
        String path;
        String iconPath;
        int ramUsageKB;
        float cpuUsage;

        public ExeProcess(String name, int unixPid, String cwd, String path, String iconPath, int ramUsageKB, float cpuUsage) {
            this.name = name;
            this.unixPid = unixPid;
            this.cwd = cwd;
            this.path = path;
            this.iconPath = iconPath;
            this.ramUsageKB = ramUsageKB;
            this.cpuUsage = cpuUsage;
        }

        public String getName() {
            return name;
        }

        public int getRamUsageKB() {
            return ramUsageKB;
        }

        public float getCpuUsage() {
            return cpuUsage;
        }

        public String getIconPath() {
            return iconPath;
        }

        public int getUnixPid() {
            return unixPid;
        }
    }
}

File: app/src/main/java/com/micewine/emu/core/WineWrapper.kt
package com.micewine.emu.core

import com.micewine.emu.activities.MainActivity.Companion.deviceArch
import com.micewine.emu.activities.MainActivity.Companion.selectedCpuAffinity
import com.micewine.emu.activities.MainActivity.Companion.usrDir
import com.micewine.emu.activities.MainActivity.Companion.wineDisksFolder
import com.micewine.emu.activities.MainActivity.Companion.winePrefix
import com.micewine.emu.activities.MainActivity.Companion.winePrefixesDir
import com.micewine.emu.core.EnvVars.getEnv
import com.micewine.emu.core.ShellLoader.runCommand
import com.micewine.emu.core.ShellLoader.runCommandWithOutput
import com.micewine.emu.fragments.DebugSettingsFragment.Companion.availableCPUs
import java.io.File
import kotlin.math.abs

object WineWrapper {
    private var IS_BOX64 = if (deviceArch == "x86_64") "" else "box64"

    fun getCpuHexMask(cpuAffinityMask: String = selectedCpuAffinity!!): String {
        val availCpus = Runtime.getRuntime().availableProcessors()
        val cpuMask = MutableList(availCpus) { '0' }
        val cpuAffinity = cpuAffinityMask.replace(",", "")

        for (element in cpuAffinity) {
            cpuMask[abs(element.toString().toInt() - availCpus) - 1] = '1'
        }

        return Integer.toHexString(cpuMask.joinToString("").toInt(2))
    }

    fun maskToCpuAffinity(mask: Long): BooleanArray {
        val availCpus = Runtime.getRuntime().availableProcessors()
        val affinity = BooleanArray(availCpus) { false }
        for (i in 0 until availCpus) {
            if ((mask shr i) and 1L == 1L) {
                affinity[i] = true
            }
        }
        return affinity
    }


    fun waitForProcess(name: String) {
        while (true) {
            val wineProcesses = getExeProcesses()
            if (wineProcesses.firstOrNull { it.name == name } != null) break
            Thread.sleep(125)
        }
    }

    fun wine(args: String) {
        runCommand(
            getEnv() + "WINEPREFIX='$winePrefixesDir/$winePrefix' $IS_BOX64 wine $args"
        )
    }

    fun wine(args: String, retLog: Boolean): String {
        if (retLog) {
            return runCommandWithOutput(
                getEnv() + "BOX64_LOG=0 WINEPREFIX='$winePrefixesDir/$winePrefix' $IS_BOX64 wine $args"
            )
        }
        return ""
    }

    fun wine(args: String, cwd: String) {
        runCommand(
            "cd $cwd;" + getEnv() + "WINEPREFIX='$winePrefixesDir/$winePrefix' $IS_BOX64 wine $args"
        )
    }

    fun killAll() {
        runCommand(getEnv() + "WINEPREFIX='$winePrefixesDir/$winePrefix' $IS_BOX64 wineserver -k", false)
        runCommand("pkill -SIGINT -f .exe")
        runCommand("pkill -SIGINT -f wineserver")
    }

    fun clearDrives() {
        var letter = 'e'

        while (letter <= 'y') {
            val disk = File("$wineDisksFolder/$letter:")
            if (disk.exists()) {
                disk.delete()
            }
            letter++
        }
    }

    fun addDrive(path: String) {
        runCommand("ln -sf $path $wineDisksFolder/${getAvailableDisks()[0]}:")
    }

    private fun getAvailableDisks(): List<String> {
        var letter = 'c'
        val availableDisks = mutableListOf<String>()

        while (letter <= 'z') {
            if (!File("$wineDisksFolder/$letter:").exists()) {
                availableDisks.add("$letter")
            }
            letter++
        }

        return availableDisks
    }

    fun extractIcon(exePath: String, output: String) {
        if (exePath.endsWith(".exe", true)) {
            runCommand(
                getEnv() + "wrestool -x -t 14 '${getSanitizedPath(exePath)}' > '$output'", false
            )
        }
    }

    fun getSanitizedPath(filePath: String) : String {
        return filePath.replace("'", "'\\''")
    }

    private fun getProcessPath(processName: String, processCwd: String): String {
        val file = File(processCwd, processName)
        if (file.exists()) return file.path

        val searchPaths = listOf(
            "$wineDisksFolder/c:/windows/system32",
            "$wineDisksFolder/c:/windows"
        )
        searchPaths.forEach { path ->
            val exeFile = File(path, processName)
            if (exeFile.exists()) return "$path/$processName"
        }

        return ""
    }

    private fun getProcessRamUsageKB(pid: Int): Int {
        val statusFile = File("/proc/$pid/status")
        if (!statusFile.exists()) return 0

        statusFile.readLines().forEach {
            if (it.startsWith("VmRSS:")) {
                return it.substringAfter(":").replace("kB", "").trim().toIntOrNull() ?: 0
            }
        }
        return 0
    }

    fun getProcessCPUAffinity(pid: Int): String {
        val statusFile = File("/proc/$pid/status")
        if (!statusFile.exists()) return "0"

        statusFile.readLines().forEach {
            if (it.startsWith("Cpus_allowed:")) {
                return it.substringAfter(":").replace(",", "").trim()
            }
        }
        return "0"
    }

    fun getExeProcesses(): List<ExeProcess> {
        val exeProcesses = mutableListOf<ExeProcess>()
        File("/proc").listFiles()?.forEach {
            val unixPid = it.name.toIntOrNull()
            if (it.isDirectory && unixPid != null) {
                val cmdlineFile = File(it, "cmdline")
                if (cmdlineFile.exists()) {
                    val cmdline = cmdlineFile.readText().trim()
                    val processName = cmdline.split("\u0000").firstOrNull()
                    if (processName != null && processName.endsWith(".exe", true)) {
                        val cwd = runCommandWithOutput("readlink $it/cwd").trim()
                        val path = getProcessPath(processName, cwd)
                        val iconPath = "$usrDir/icons/${processName.substringBefore(".exe")}-icon"
                        val ramUsageKB = getProcessRamUsageKB(unixPid)
                        val cpuUsage = runCommandWithOutput("ps -p $unixPid -o %cpu=").trim().toFloatOrNull() ?: 0F

                        extractIcon(path, iconPath)

                        exeProcesses.add(
                            ExeProcess(processName, unixPid, cwd, path, iconPath, ramUsageKB, cpuUsage / availableCPUs.size)
                        )
                    }
                }
            }
        }
        return exeProcesses
    }

    class ExeProcess(
        val name: String,
        val unixPid: Int,
        val cwd: String,
        val path: String,
        val iconPath: String,
        val ramUsageKB: Int,
        val cpuUsage: Float
    )
}